src/libcore/str/mod.rs

```diff
@@ -1741,13 +1741,29 @@ fn truncate_to_char_boundary(s: &str, mut max: usize) -> (bool, &str) {
 #[cold]
 fn slice_error_fail(s: &str, begin: usize, end: usize) -> ! {
     const MAX_DISPLAY_LENGTH: usize = 256;
-    let (truncated, s) = truncate_to_char_boundary(s, MAX_DISPLAY_LENGTH);
+    let (truncated, s_trunc) = truncate_to_char_boundary(s, MAX_DISPLAY_LENGTH);
     let ellipsis = if truncated { "[...]" } else { "" };
 
+    // 1. out of bounds
+    let index = if !s.is_char_boundary(begin) { begin } else { end };
```

Comment:
This is confusing to me. It seems like you choose which index is going to be in the message before checking which one is out of bounds.